File: daily/rust/39_combination_sum.rs
impl Solution {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut res = vec![];
        let mut path = vec![];
        Self::dfs(&candidates, target, 0, &mut path, &mut res);
        res
    }
    
    fn dfs(candidates: &Vec<i32>, target: i32, start: usize, path: &mut Vec<i32>, res: &mut Vec<Vec<i32>>) {
        if target < 0 {
            return;
        }
        if target == 0 {
            res.push(path.clone());
            return;
        }
        for i in start..candidates.len() {
            path.push(candidates[i]);
            Self::dfs(candidates, target - candidates[i], i, path, res);
            path.pop();
        }
    }
}

File: daily/cpp/39_combination_sum.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  /* A typical backtrack solution */
  vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    vector<vector<int>> res;
    vector<int> path;
    backtrack(candidates, target, 0, path, res);
    return res;
  }

 private:
  void backtrack(vector<int>& candidates, int target, int start,
                 vector<int>& path, vector<vector<int>>& res) {
    if (target < 0) return;
    if (target == 0) {
      res.push_back(path);
      return;
    }
    for (int i = start; i < candidates.size(); ++i) {
      path.push_back(candidates[i]);
      backtrack(candidates, target - candidates[i], i, path, res);
      path.pop_back();
    }
  }
};